pkg/bpf/map_linux.go

```diff
@@ -173,15 +173,21 @@ func NewMap(name string, mapType ebpf.MapType, mapKey MapKey, mapValue MapValue,
 	keySize := reflect.TypeOf(mapKey).Elem().Size()
 	valueSize := reflect.TypeOf(mapValue).Elem().Size()
 
+	return NewMapWithSpec(name, mapKey, mapValue, &ebpf.MapSpec{
+		Type:       mapType,
+		Name:       path.Base(name),
+		KeySize:    uint32(keySize),
+		ValueSize:  uint32(valueSize),
+		MaxEntries: uint32(maxEntries),
+		Flags:      flags,
+		Pinning:    ebpf.PinByName,
+	})
+}
+
+// NewMapWithSpec creates a new Map instance - object representing a BPF map
+func NewMapWithSpec(name string, mapKey MapKey, mapValue MapValue, spec *ebpf.MapSpec) *Map {
```

Comment:
Let's try to avoid adding yet another way of creating a map. Could you elaborate a bit on why this change is necessary? Currently, pinning behaviour is governed completely by OpenOrCreate(), Recreate() and CreateUnpinned(). With your patch, you can enable pinning through `NewMapWithSpec`, but a call to `CreateUnpinned()` will mutate the spec and create it without pinning regardless. This will make our planned refactoring a bit painful and error-prone.

To be clear, if this API currently looks like a mess to you, that's because it is. :wink: My goal was for `NewMap` to eventually take a MapSpec, so I agree with your approach of introducing `NewMapWithSpec`, but that means we also need to get rid of `CreateUnpinned()` and `OpenOrCreate()` at the same time, which is a bigger challenge (see https://github.com/cilium/cilium/issues/28742, https://github.com/cilium/cilium/issues/25224 and https://github.com/cilium/cilium/issues/26793).

Would this PR work without the `NewMapWithSpec` changes as well? :pray:

Reply:
Oh, I didn't know you were already planning those refactors.

> Could you elaborate a bit on why this change is necessary?

I thought  `CreateUnPInned` is going to be removed and shouldn't be used from this comment. So I added `NewMapWithSpec`. (I was about to migrate to NewMapWithSpec as a follow-on)
https://github.com/cilium/cilium/blob/main/pkg/bpf/map_linux.go#L466

> Would this PR work without the NewMapWithSpec changes as well?

I think we can do it using `CreateUnpinned`. If a caller passes `PinNone` then call `CreateUnpinned`, `PinByName` then call `OpenOrCreate`.

Reply:
> I thought CreateUnPInned is going to be removed and shouldn't be used from this comment.

Oh, I see, sorry for the confusion. :sweat_smile: If CreateUnpinned() was to be avoided, I would've put a `Deprecated:` comment there that gets flagged by Go tooling. Feel free to use it for now, I prefer if the intent is clear so we can easily pick this up during a refactor. :+1:

Reply:
I dropped `NewMapWithSpec`. Please take a look